lineapy/plugins/pipeline_writers.py

```diff
@@ -172,44 +176,79 @@ def _write_dag(self) -> None:
+                pickle.dump(artifacts["z"], open("/tmp/g2_z/artifact_z.pickle", "wb"))
 
             with DAG(...) as dag:
-                run_session_including_iris_model = PythonOperator(
-                    task_id="run_session_including_iris_model_task",
-                    python_callable=iris_module.run_session_including_iris_model,
+                setup = PythonOperator(
+                    task_id="dag_setup",
+                    python_callable=dag_setup,
+                )
+
+                teardown = PythonOperator(
+                    task_id="dag_teardown",
+                    python_callable=dag_teardown,
                 )
 
-        This makes the Airflow DAG file more compact but lack transparency
-        and fuller control over each artifact-level task. Check the
-        `PythonOperatorPerArtifact` flavor to address this limitation.
+                run_session_including_g2 = PythonOperator(
+                    task_id="run_session_including_g2_task",
+                    python_callable=task_run_session_including_g2,
+                )
+
+                setup >> run_session_including_g2
+                run_session_including_g2 >> teardown
+
+            run_session_including_g2
+
+        This way, the generated Airflow DAG file opens room for engineers
+        to peak and control pipeline runs at a finer level and allows
+        for further customization.
         """
-        DAG_TEMPLATE = load_plugin_template(
-            "airflow_dag_PythonOperatorPerSession.jinja"
-        )
-        session_functions = [
-            f"run_session_including_{session_artifacts._get_first_artifact_name()}"
-            for session_artifacts in self.session_artifacts_sorted
-        ]
+        codegenerator = AirflowCodeGenerator(self.artifact_collection)
+        DAG_TEMPLATE = load_plugin_template("airflow_dag_PythonOperator.jinja")
+        task_functions = []
+        task_definitions = []
+        for sa in self.session_artifacts_sorted:
+            task_functions.append(sa.get_session_function_name())
+            task_definitions.append(
+                get_session_task_definition(sa, self.pipeline_name)
+            )
         dependencies = {
-            session_functions[i + 1]: {session_functions[i]}
-            for i in range(len(session_functions) - 1)
+            task_functions[i + 1]: {task_functions[i]}
+            for i in range(len(task_functions) - 1)
         }
         task_graph = TaskGraph(
-            nodes=session_functions,
-            mapping={f: f for f in session_functions},
+            nodes=task_functions,
+            mapping={f: f for f in task_functions},
             edges=dependencies,
         )
+        session_function_params = (
+            codegenerator.get_session_function_params_args()
+        )
+        tasks = [
+            (ft, session_function_params.get(ft, None))
+            for ft in task_functions
+        ]
```

Comment:
Can we make `tasks` a dictionary instead? This way, we make the corresponding template (i.e., `airflow_dag_PythonOperator.jinja`) clearer. Specifically, the template currently has it as the following:

```jinja
{% for TASKDESCRIPTION in tasks %}
{% if TASKDESCRIPTION[1] is none %}
    {{ TASKDESCRIPTION[0] }} = PythonOperator(
        task_id="{{ TASKDESCRIPTION[0] }}_task", python_callable=task_{{ TASKDESCRIPTION[0] }},
    )
{% else %}
    {{ TASKDESCRIPTION[0] }} = PythonOperator(
        task_id="{{ TASKDESCRIPTION[0] }}_task", python_callable=task_{{ TASKDESCRIPTION[0] }}, {{TASKDESCRIPTION[1]}},
    )
{% endif %}
{% endfor %}
```

where we rely on numbered index (`TASKDESCRIPTION[0]`) which is not clear as to what it contains. Instead having `tasks` as a dictionary will help to clarify this as we then will be using `TASKDESCRIPTION.[informative-key-name]` in the template.